plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/tracing/TracingHiveMetastore.java

```diff
@@ -0,0 +1,712 @@
+import io.trino.plugin.hive.metastore.Partition;
+import io.trino.plugin.hive.metastore.PartitionWithStatistics;
+import io.trino.plugin.hive.metastore.PrincipalPrivileges;
+import io.trino.plugin.hive.metastore.Table;
+import io.trino.spi.connector.SchemaTableName;
+import io.trino.spi.function.LanguageFunction;
+import io.trino.spi.predicate.TupleDomain;
+import io.trino.spi.security.RoleGrant;
+import io.trino.spi.type.Type;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.OptionalLong;
+import java.util.Set;
+import java.util.function.Function;
+
+import static io.opentelemetry.api.common.AttributeKey.stringKey;
+import static java.util.Objects.requireNonNull;
+
+public class TracingHiveMetastore
+        implements HiveMetastore
+{
+    private static final AttributeKey<String> SCHEMA = stringKey("trino.schema");
+    private static final AttributeKey<String> TABLE = stringKey("trino.table");
+    private static final AttributeKey<String> PARTITION = stringKey("trino.partition");
+    private static final AttributeKey<String> FUNCTION = stringKey("trino.function");
+    private static final AttributeKey<String> ACID_TRANSACTION = stringKey("trino.acid_transaction");
+
+    private final Tracer tracer;
+    private final HiveMetastore delegate;
+
+    public TracingHiveMetastore(Tracer tracer, HiveMetastore delegate)
+    {
+        this.tracer = requireNonNull(tracer, "tracer is null");
+        this.delegate = requireNonNull(delegate, "delegate is null");
+    }
+
+    @Override
+    public Optional<Database> getDatabase(String databaseName)
+    {
+        Span span = startSpan("getConnectorCapabilities", databaseName);
+        try (var ignored = scopedSpan(span)) {
+            return delegate.getDatabase(databaseName);
+        }
+    }
+
+    @Override
+    public List<String> getAllDatabases()
+    {
+        Span span = startSpan("getAllDatabases");
+        try (var ignored = scopedSpan(span)) {
+            return delegate.getAllDatabases();
+        }
+    }
+
+    @Override
+    public Optional<Table> getTable(String databaseName, String tableName)
+    {
+        Span span = startSpan("getTable", databaseName, tableName);
+        try (var ignored = scopedSpan(span)) {
+            return delegate.getTable(databaseName, tableName);
+        }
+    }
+
+    @Override
+    public Set<HiveColumnStatisticType> getSupportedColumnStatistics(Type type)
+    {
+        // Tracing is not necessary
+        return delegate.getSupportedColumnStatistics(type);
+    }
+
+    @Override
+    public PartitionStatistics getTableStatistics(Table table)
+    {
+        Span span = startSpan("getTableStatistics", table);
+        try (var ignored = scopedSpan(span)) {
+            return delegate.getTableStatistics(table);
+        }
+    }
+
+    @Override
+    public Map<String, PartitionStatistics> getPartitionStatistics(Table table, List<Partition> partitions)
```

Comment:
We could add a `trino.hive.request.partition_count` and `trino.hive.response.partition_count` attributes